Add onboarding payment gateway setup methods

joelclimbsthings opened the pull request:
Part of - https://github.com/woocommerce/woocommerce-admin/issues/6878

Adds the methods needed to use the config-based setup for the payment gateway during onboarding.

![image](https://user-images.githubusercontent.com/444632/120547290-bfcf8300-c3a5-11eb-8aab-186ee946c37f.png)


#### Testing instructions
1. Install and activate this plugin in your `plugins/` directory.
2. Install and activate the `add/remote-payment-wcpay` branch of [WooCommerce Admin](https://github.com/woocommerce/woocommerce-admin) ( `git clone https://github.com/woocommerce/woocommerce-admin.git && cd woocommerce-admin  && git checkout add/remote-payment-wcpay && composer install && npm i && npm run build` ).
3. Make sure a qualifying WCPay is selected, and that the "cbd" industry is not selected during onboarding.
4. Navigate to the payment tasks via _WooCommerce -> Home -> Set up payments_
5. Click "Finish Setup" on WCPay card.
6. Make sure that the description text is displayed at the connect button directs you to the correct place.

-------------------

- [x] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
-->

- [ ] Added testing instructions to the [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) (or does not apply)

joshuatf commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -2125,4 +2125,22 @@ public function get_upe_available_payment_methods() {
+	 *
+	 * @return string Connection URL.
+	 */
+	public function get_oauth_connection_url() {
```
Just a heads up that this may be updated to `get_connection_url()` in https://github.com/woocommerce/woocommerce-admin/pull/7095 and `$return_url` will be passed here for use.

joelclimbsthings replied:
Updated 👍🏻

joshuatf commented on includes/class-wc-payments-account.php:
```diff
@@ -394,7 +394,15 @@ public static function get_login_url() {
+			admin_url( 'admin.php' )
+		);
+
+		return wp_nonce_url( add_query_arg( [ 'wcpay-connect' => '1' ], $connect_url ), 'wcpay-connect' );
```
I think `wcpay-connect` may no longer be used.

joelclimbsthings replied:
Are you sure? I'm just mimicking the URL that was already in use here.

htdat replied:
At first, I don't know why we need to make this change. However, it makes sense when I test this PR. Otherwise, the `Connect` button's link here will be something like http://localhost:8082/wp-json/wc/v3/payment_gateways?_locale=user&wcpay-connect=1&_wpnonce=xxxxxxx 


> It actually looks like this is needed to initiate the oauth redirect to Stripe, which is what I intended. This is handled here.
Write

That's right, we should still keep `wcpay-connect` as WCPay uses it to generate the Stripe onboarding link.

However, we do not need to build up $connection_url like this. Instead, using `admin_url( 'admin.php' )` is quite enough and avoids side effects by introducing more parameters.

Note: the below suggestion is for the whole change of this `get_connect_url` method. 

```suggestion
		return wp_nonce_url( add_query_arg( [ 'wcpay-connect' => '1' ], admin_url( 'admin.php' ) ), 'wcpay-connect' );
```

joshuatf commented on includes/class-wc-payments-account.php:
```diff
@@ -394,7 +394,15 @@ public static function get_login_url() {
+		$connect_url = add_query_arg(
+			[
+				'page' => 'wc-admin',
+				'path' => '/payments/connect',
```
Is this a route added somewhere or just a placeholder?  Some updates to returned connections here - https://github.com/woocommerce/woocommerce-admin/pull/7095

joelclimbsthings replied:
This directs to the primary connect page, but the `wcpay-connect` parameter then redirects to the onboarding flow with Stripe.

htdat commented:
> 2. Install and activate the `add/remote-payment-wcpay` branch of [WooCommerce Admin](https://github.com/woocommerce/woocommerce-admin) ( `git clone https://github.com/woocommerce/woocommerce-admin.git && cd woocommerce-admin  && git checkout add/remote-payment-wcpay && composer install && npm i && npm run build` ).

@joshuatf and @joelclimbsthings This instruction step is still correct? I am not able to find [`add/remote-payment-wcpay` branch](https://github.com/woocommerce/woocommerce-admin/branches/all?query=remote-payment-wcpay) on the woocommerce-admin repo.

Also FYI. I've updated this PR's description with a few checklist items. Don't worry much about that in case you're not familiar with them. I can assist to finish them if applicable.

htdat reviewed (changes requested):
OK, I can test this PR with WooCommerce 5.9.0 without the need of `woocommerce-admin`. I guess that's because the necessary change has been merged into the Woo core. 

Notes:

- I needed to deactivate WCPay and ensure that I did not go through the WCPay onboarding for my test site.
- This bug can happen in https://github.com/Automattic/woocommerce-payments/issues/3333 but to test this PR, just deactivating WCPay is enough. 

> 5. Click "Finish Setup" on WCPay card.

To be clear, I confirm to see this but on this page `wp-admin/admin.php?page=wc-admin&task=payments`, which might be confusing with WCPay connect page at `/wp-admin/admin.php?page=wc-admin&path=%2Fpayments%2Fconnect`

✅ Following test instructions with the notes above, and confirm this PR works as expected. 
✅ Generally, changes on this PR looks good.
❓ However, I would minimize the change for this method `get_connect_url`. More details in my inline comment.

joelclimbsthings commented:
Thanks for the feedback @htdat , much appreciated! I've gone ahead and committed your suggestions regarding the connect URL.

> This instruction step is still correct? I am not able to find add/remote-payment-wcpay branch on the woocommerce-admin repo.

As you guessed, the changes in that PR were merged long ago. Just using the `main` branch of WCAdmin, or, it seems, the latest Woo Core, should work just fine. 

I'm going to update this on the base branch again. Any further feedback or guidance on how to get this over the line would be appreciated as well.

htdat commented on includes/class-wc-payments-account.php:
```diff
@@ -419,7 +419,15 @@ private function get_login_url() {
 	 */
 	public static function get_connect_url() {
-		return wp_nonce_url( add_query_arg( [ 'wcpay-connect' => '1' ] ), 'wcpay-connect' );
+		$connect_url = add_query_arg(
```
Should we remove this var `$connect_url` too as we're not using it anymore?

htdat commented on includes/class-wc-payments-account.php:
```diff
@@ -419,7 +419,15 @@ private function get_login_url() {
 	 */
 	public static function get_connect_url() {
-		return wp_nonce_url( add_query_arg( [ 'wcpay-connect' => '1' ] ), 'wcpay-connect' );
+		$connect_url = add_query_arg(
```
Should we remove this var `$connect_url` too as we're not using it anymore?

joelclimbsthings replied:
Definitely 👍🏻

htdat reviewed (changes requested):
@joelclimbsthings Thanks for updating the PR. 

A new conflict has just happened. Can you resolve it?
Also, it's great if you can add changelog entries for [changelog.txt](https://github.com/Automattic/woocommerce-payments/blob/400142062aeb835faf40befcdefa8fbc49a5c29b/changelog.txt#L12-L12) and [readmet.txt](https://github.com/Automattic/woocommerce-payments/blob/400142062aeb835faf40befcdefa8fbc49a5c29b/readme.txt#L109-L109)

joelclimbsthings commented:
Thanks again @htdat . I've removed the orphaned variable, resolved the conflict, and added the changelog entries. 

Just let me know if anything else looks amiss.

htdat commented:
Thanks @joelclimbsthings! 

> Just let me know if anything else looks amiss.

It seems that the following change is still needed. Otherwise, merchants will be routed to an unexpected link http://localhost:8082/wp-json/wc/v3/payment_gateways?_locale=user&wcpay-connect=1&_wpnonce=xxxxxxx

```diff
Index: includes/class-wc-payments-account.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/includes/class-wc-payments-account.php b/includes/class-wc-payments-account.php
--- a/includes/class-wc-payments-account.php	(revision d4cf779346514e90415fe0d6711393412c195c24)
+++ b/includes/class-wc-payments-account.php	(date 1638236778412)
@@ -422,7 +422,7 @@
 	 * @return string Stripe account login url.
 	 */
 	public static function get_connect_url( $wcpay_connect_from = '1' ) {
-		return wp_nonce_url( add_query_arg( [ 'wcpay-connect' => $wcpay_connect_from ] ), 'wcpay-connect' );
+		return wp_nonce_url( add_query_arg( [ 'wcpay-connect' => $wcpay_connect_from ], admin_url( 'admin.php' ) ), 'wcpay-connect' );
 	}
 
 	/**
```

htdat commented on changelog.txt:
```diff
@@ -9,6 +9,7 @@
 * Fix - Updates to fraud protection.
 * Fix - Prevent Payment Request buttons from showing up in Composite Product pages.
 * Update - Updated @woocommerce/experimental package to v2.1.0.
+* Add - Add support for suggested gateway methods in WC-Admin
```
Nit: may add the dot at the end of this entry? Similar to the `readme.txt` file.

joelclimbsthings commented:
Thanks again @htdat 

> It seems that the following change is still needed. Otherwise, merchants will be routed to an unexpected link http://localhost:8082/wp-json/wc/v3/payment_gateways?_locale=user&wcpay-connect=1&_wpnonce=xxxxxxx

Ah, looks like I lost that during the conflict resolution. Nice catch!

Fixed the readme's as well. Appreciate the feedback, and let me know if anything else needs addressing.

htdat reviewed (approved):
LGTM! 🚢

htdat commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -2656,4 +2656,22 @@ public function get_upe_available_payment_methods() {
+	}
+
+	/**
+	 * Get the oAuth connection URL.
```
Thanks @dechov! Yes, I am going to update it in PR #3496
